Composer/packages/client/src/recoilModel/dispatchers/auth.ts

```diff
@@ -0,0 +1,156 @@
+/* eslint-disable react-hooks/rules-of-hooks */
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+import formatMessage from 'format-message';
+import { CallbackInterface, useRecoilCallback } from 'recoil';
+import { AzureTenant } from '@botframework-composer/types';
+
+import {
+  userShouldProvideTokens,
+  isShowAuthDialog,
+  getTokenFromCache,
+  setTenantId,
+  getTenantIdFromCache,
+} from '../../utils/auth';
+import {
+  showAuthDialogState,
+  primaryTokenState,
+  graphTokenState,
+  availableTenantsState,
+  showTenantDialogState,
+  currentTenantState,
+} from '../atoms/appState';
+import { AuthClient } from '../../utils/authClient';
+import storage from '../../utils/storage';
+
+import { addNotificationInternal, createNotification } from './notification';
+
+export const authDispatcher = () => {
+  const setShowAuthDialog = useRecoilCallback((callbackHelpers: CallbackInterface) => (show: boolean) => {
+    callbackHelpers.set(showAuthDialogState, show);
+  });
+
+  const setPrimaryToken = useRecoilCallback((callbackHelpers: CallbackInterface) => (token: string) => {
+    callbackHelpers.set(primaryTokenState, token);
+    storage.set('accessToken', token);
+  });
+
+  const setGraphToken = useRecoilCallback((callbackHelpers: CallbackInterface) => (token: string) => {
+    callbackHelpers.set(graphTokenState, token);
+    storage.set('graphToken', token);
+  });
+
+  const setCurrentTenant = useRecoilCallback(
+    (callbackHelpers: CallbackInterface) => async (tenant: string, notify = true) => {
+      callbackHelpers.set(currentTenantState, tenant);
+      setTenantId(tenant);
+      if (tenant) {
+        // get arm token for tenant
+        try {
+          const token = await AuthClient.getARMTokenForTenant(tenant);
+          if (token) {
+            setPrimaryToken(token);
+
+            // fire notification
+            if (notify !== false) {
+              // set notification
+              const notification = createNotification({
+                title: formatMessage('Azure sign-in'),
+                description: formatMessage("You've successfully signed in."),
+                type: 'info',
+              });
+              addNotificationInternal(callbackHelpers, notification);
+            }
+          }
+        } catch (err) {
+          const notification = createNotification({
+            title: formatMessage('Azure sign-in'),
+            description: formatMessage(`Sign in failed: {message}`, { message: err.message || err.toString() }),
+            type: 'error',
+          });
+          addNotificationInternal(callbackHelpers, notification);
+          // clear out app state
+          resetCreds();
+        }
+      }
+    }
+  );
+
+  const setAvailableTenants = useRecoilCallback((callbackHelpers: CallbackInterface) => (tenants: AzureTenant[]) => {
```

Comment:
Nit: You don't have to put this in the dispatcher unless you see logic being added here later, components can use recoil directly.